cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -640,7 +641,18 @@ public void deleteApplication(String namespaceId, String appId, String versionId
     getApplicationSpecificationTable().delete(fields);
   }
 
-  public void deleteApplicationEditRecord(ApplicationReference appRef) throws IOException {
+  private void deleteApplication(ApplicationReference appRef)
+    throws IOException {
+    getApplicationSpecificationTable().deleteAll(getNamespaceApplicationAndStatusRange(appRef));
+  }
+
+  public void removeApplicationData(ApplicationReference appRef) throws IOException {
+    deleteApplication(appRef);
+    deleteApplicationEditRecord(appRef);
+    deleteProgramHistory(appRef);
```

Comment:
`deleteApplicationEditRecord` can be deleted initially. However, `deleteProgramHistory` is across all versions, is it okay to delete it synchronously? I agree that it would create a race condition otherwise

Reply:
That's a good call. If we assume someone can reinstate the app, are we going to delete anything  when we process the delete? I mean, this just blocks us from deleting anything in async fashion. Should we add a timestamp to the messages? Otherwise we won't be able to process anything asynchronously. I am kind of OK returning "ghosts" for resurrected apps, but I am not OK for deleting newer data

Reply:
timestamp seems like a good idea, we can delete all program history before that timestamp. It's possible to fix the 'ghost' edge case if we also change the run record list to get only runs started after the first version of the app, but I don't think we have a performant way to get the creation time of the first version?

I think the application edit record still needs to be deleted with the initial marking though.